File: 백준/Silver/9012. 괄호/괄호.cc
#include <iostream>
#include <stack>
using namespace std;
int main() {
	ios::sync_with_stdio(false); cin.tie(0); cout.tie(0);
	int T;
	string s;
	cin >> T;
	for (int i = 0; i < T; i++)
	{
		stack <char> vps;
		cin >> s;
		for (int j = 0; j < s.length(); j++)
		{
			vps.push(s[j]);
		}
		int cnt = 0;
		int size = vps.size();
		int k = 0;
		for (k = 0; k < size; k++)
		{
			if (vps.top() == '(') {
				if (cnt > 0) {
					cnt--;
					vps.pop();
				}
				else {
					cout << "NO" << '\n';
					break;
				}
			}
			else if (vps.top() == ')')
			{
				if (cnt < 0)
				{
					cout << "NO" << '\n';
					break;
				}
				else {
					cnt++;
					vps.pop();
				}

			}
		}
		if (k == size)
		{
			if (cnt > 0)
			{
				cout << "NO" << '\n';
			}
			else {
				cout << "YES" << '\n';
			}
		}
	}
	return 0;
}

File: 백준/Silver/9012. 괄호/괄호.py
import sys
input = sys.stdin.readline

for _ in range(int(input())):
    stack = []
    n = list(input().rstrip())
    for i in n:
        if i == ')':
            if len(stack) != 0 and stack[-1] == '(':
                stack.pop()
            else :
                stack.append(')')
        elif i == '(':
            stack.append(i)
    if len(stack) != 0:
        print('NO')
    else :
        print('YES')